File: client/python/modeldb/events/RandomSplitEvent.py
"""
Event indicating that a dataFrame was split into smaller dataFrames.
"""
from modeldb.events.Event import Event
from ..thrift.modeldb import ttypes as modeldb_types


class RandomSplitEvent(Event):
    """
    Class for creating and storing RandomSplitEvents
    """

    def __init__(self, df, weights, seed, result):
        self.df = df
        self.weights = weights
        self.seed = seed
        self.result = result

    def make_event(self, syncer):
        """
        Constructs a thrift RandomSplitEvent object with appropriate fields.
        """
        syncable_dataframe = syncer.convert_df_to_thrift(self.df)
        all_syncable_frames = []
        for dataframe in self.result:
            all_syncable_frames.append(syncer.convert_df_to_thrift(dataframe))
        re = modeldb_types.RandomSplitEvent(
            syncable_dataframe, self.weights, self.seed, all_syncable_frames,
            syncer.experiment_run.id)
        return re

    def associate(self, res, syncer):
        """
        Stores the server response ids for all split dataframes into
        dictionary.
        """
        syncer.store_object(self.df, res.oldDataFrameId)
        for i in range(0, len(self.result)):
            syncer.store_object(self.result[i], res.splitIds[i])
        syncer.store_object(self, res.splitEventId)

    def sync(self, syncer):
        """
        Stores RandomSplitEvent on the server.
        """
        re = self.make_event(syncer)
        thrift_client = syncer.client
        res = thrift_client.storeRandomSplitEvent(re)
        self.associate(res, syncer)

File: client/scala/libs/spark.ml/src/main/scala-2.11/edu/mit/csail/db/ml/modeldb/client/event/RandomSplitEvent.scala
package edu.mit.csail.db.ml.modeldb.client.event

import com.twitter.util.Await
import edu.mit.csail.db.ml.modeldb.client.{ModelDbSyncer, SyncableDataFrame}
import modeldb.ModelDBService.FutureIface
import org.apache.spark.sql.DataFrame

/**
  * Event indicating that a DataFrame was split into smaller DataFrames.
  *
  * @param dataframe - The DataFrame that was split.
  * @param weights - The weights used to split the DataFrame into
  *                smaller pieces. For example: Array(0.5, 0.5, 1) splits
  *                the DataFrame into three pieces, where the first two
  *                pieces are half as large as the third piece.
  * @param seed - The seed for the random number generator used for the splitting.
  * @param result - The pieces produced by the splitting process. There should
  *               be one entry here for each entry in weights.
  */
case class RandomSplitEvent(dataframe: DataFrame,
                            weights: Array[Double],
                            seed: Long,
                            result: Array[DataFrame]) extends ModelDbEvent {
  def makeEvent(mdbs: ModelDbSyncer): modeldb.RandomSplitEvent = {
    modeldb.RandomSplitEvent(
      SyncableDataFrame(dataframe),
      weights,
      seed,
      result.map(df => SyncableDataFrame(df)),
      experimentRunId = mdbs.experimentRun.id
    )
  }

  def associate(res: modeldb.RandomSplitEventResponse, mdbs: ModelDbSyncer) = {
    mdbs.associateObjectAndId(dataframe, res.oldDataFrameId)
    result.zipWithIndex.foreach { case (df, i) =>
      mdbs.associateObjectAndId(result(i), res.splitIds(i))
    }
    mdbs.associateObjectAndId(this, res.splitEventId)
  }

  override def sync(client: FutureIface, mdbs: Option[ModelDbSyncer]): Unit = {
    val event = makeEvent(mdbs.get)
    val res = Await.result(client.storeRandomSplitEvent(event))
    associate(res, mdbs.get)
  }
}